R/expect-self-test.R

```diff
@@ -1,30 +1,43 @@
-capture_failure <- new_capture("expectation_failure")
-capture_success <- function(expr) {
+capture_success_failure <- function(expr) {
   cnd <- NULL
 
-  withCallingHandlers(
-    expr,
-    expectation_failure = function(cnd) {
-      invokeRestart("continue_test")
-    },
-    expectation_success = function(cnd) {
-      cnd <<- cnd
-    }
+  n_success <- 0
+  n_failure <- 0
+
+  last_failure <- NULL
+
+  withRestarts(
+    withCallingHandlers(
+      expr,
+      expectation_failure = function(cnd) {
+        last_failure <<- cnd
+        n_failure <<- n_failure + 1
+        # Finish the test without bubbling up
+        invokeRestart("failed")
```

Comment:
IIUC `n_failure` is either 0 or 1, because we're unwinding back to `failed` at the first failure?

In this case I think it would be a little clearer to remove `n_failure` and rename `last_failure` (whose "last" prefix imply there might be more than one assigned here) to just `failure`. Then you test for `!is.null(failure)` instead of `n_failure > 0`.

Also fine as is but then I'd add a comment saying we expect at most 1 failure.

Reply:
I think this is a good suggestion but I'm going to leave it until after #2129 because there I am encountering a weird situation where `fail()` is not causing an expectation to exit and I think that this might help me debug it.